streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -303,76 +271,166 @@ private static Bytes getKey(final Bytes keyBytes) {
 
     }
 
-    private static class InMemoryWindowStoreIterator implements WindowStoreIterator<byte[]> {
+    private abstract class InMemoryWindowStoreIteratorWrapper implements Comparable<InMemoryWindowStoreIteratorWrapper> {
+
+        private Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator;
+        private Iterator<Map.Entry<Bytes, byte[]>> recordIterator;
+        private KeyValue<Bytes, byte[]> next;
+        private long currentTime;
 
-        private ListIterator<KeyValue<Long, byte[]>> iterator;
+        private final boolean allKeys;
+        private Bytes keyFrom;
+        private Bytes keyTo;
 
-        InMemoryWindowStoreIterator(final ListIterator<KeyValue<Long, byte[]>> iterator) {
-            this.iterator = iterator;
+        // Default constructor sets up a dummy iterator when no results are returned (eg entire fetch range is expired)
+        InMemoryWindowStoreIteratorWrapper() {
+            this.allKeys = false;
+            recordIterator = null;
+        }
+
+        InMemoryWindowStoreIteratorWrapper(final Bytes keyFrom,
+                                           final Bytes keyTo,
+                                           final Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator) {
+            this.allKeys = keyFrom == null;
```

Comment:
To be more strict, should this be `keyFrom == null && keyTo == null`? I know that today they either are both null or neither are null, but it is future-risk vulnerable if we ever change the apis to allow null values on one side.

Reply:
Fair point, will update